### PR TITLE
feat: Generate isr before begin block

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
Closes: #914

### CODE REVIEW

**Review by nashqueue — DISMISSED** [review]

**Comment by nashqueue** [comment]
Another way to do that could be to look at the header of the last block and get the app hash from it.

File: state/executor.go [hunk 1]
```diff
@@ -345,6 +345,14 @@ func (e *BlockExecutor) execute(ctx context.Context, state types.State, block *t
 		}
 	})
 
+	if e.fraudProofsEnabled {
+		isr, err := e.getAppHash()
+		if err != nil {
+			return nil, err
+		}
+		ISRs = append(ISRs, isr)
+	}
```
Comment (gupadhyaya) [inline state/executor.go:354]: do you not need to increment `currentIsrIndex`?

Comment (Manav-Aggarwal) [reply state/executor.go:354]: yes, great catch

**Comment by Manav-Aggarwal** [comment]
> Another way to do that could be to look at the header of the last block and get the app hash from it.

yes but then we need a special case for the first block so this is more versatile.

**Review by gupadhyaya — DISMISSED** [review]

File: state/executor.go [hunk 2]
```diff
@@ -322,7 +322,7 @@ func (e *BlockExecutor) execute(ctx context.Context, state types.State, block *t
 	currentIsrIndex := 0
 
 	if e.fraudProofsEnabled && currentIsrs != nil {
-		expectedLength := len(block.Data.Txs) + 2
+		expectedLength := len(block.Data.Txs) + 3
 		// BeginBlock + DeliverTxs + EndBlock
```
Comment (tzdybal) [inline state/executor.go:326]:
```suggestion
		expectedLength := len(block.Data.Txs) + 3 // before BeginBlock, after BeginBlock, after every Tx, after EndBlock
```

**Review by tzdybal — DISMISSED** [review]
Left just a small suggestion :+1:

**Review by nashqueue — APPROVED** [review]

**Review by gupadhyaya — APPROVED** [review]

**Pull request merged**